File: Programmers/[60059] 자물쇠와 열쇠.py
def rotate_matrix(m):
    # * rotate ->
    ml = len(m)
    rotated = [[0] * ml for _ in range(ml)]
    for i in range(ml):
        for j in range(ml):
            rotated[i][j] = m[ml - j - 1][i]
    return rotated


def add_matrix(lock, key, y, x):
    temp = list(map(lambda x: x[:], lock))
    for i in range(len(key)):
        for j in range(len(key)):
            temp[y + i][x + j] += key[i][j]
    return temp


def check_matrix(m, N):
    for i in range(N, 2*N):
        for j in range(N, 2*N):
            if m[i][j] != 1:
                return False
    return True


def solution(key, lock):
    N = len(lock)

    # * lock 확장 3N x 3N
    lock_extended = [[0] * 3 * N for _ in range(3*N)]
    for i in range(N, 2*N):
        for j in range(N, 2*N):
            lock_extended[i][j] = lock[i-N][j-N]

    # * key 회전시키고 이동하며 lock 홈에 다 넣어보기
    rotated_key = key
    for _ in range(4):
        for y in range(2*N):
            for x in range(2*N):
                # * N~2N 범위 다 1인지 확인
                if check_matrix(add_matrix(lock_extended, rotated_key, y, x), N):
                    return True
        rotated_key = rotate_matrix(rotated_key)

    return False


if __name__ == "__main__":
    key, lock = [[0, 0, 0], [1, 0, 0], [0, 1, 1]], [
        [1, 1, 1], [1, 1, 0], [1, 0, 1]]
    print(solution(key, lock))

File: Programmers/[60059] 자물쇠와 열쇠.java
package pg60059;

class Solution {

    static int[][] extendedLock;
    static int keySize;
    static int lockSize;

    static int[][] rotate(int[][] key) {
        int[][] rotatedKey = new int[keySize][keySize];
        for (int i = 0; i < keySize; i++) {
            for (int j = 0; j < keySize; j++) {
                rotatedKey[j][keySize - i - 1] = key[i][j];
            }
        }

        return rotatedKey;
    }

    static int[][] combine(int[][] key, int startY, int startX) {
        int[][] combinedExtendedLock = new int[lockSize + keySize * 2][lockSize + keySize * 2];
        for (int i = 0; i < lockSize + keySize * 2; i++) {
            for (int j = 0; j < lockSize + keySize * 2; j++) {
                combinedExtendedLock[i][j] = extendedLock[i][j];
            }
        }

        for (int i = 0; i < keySize; i++) {
            for (int j = 0; j < keySize; j++) {
                combinedExtendedLock[i + startY][j + startX] += key[i][j];
            }
        }

        return combinedExtendedLock;
    }

    static boolean isRightKey(int[][] combinedExtendedLock) {
        for (int i = keySize; i < keySize + lockSize; i++) {
            for (int j = keySize; j < keySize + lockSize; j++) {
                if (combinedExtendedLock[i][j] != 1) {
                    return false;
                }
            }
        }
        return true;
    }

    public boolean solution(int[][] key, int[][] lock) {

        keySize = key.length;
        lockSize = lock.length;

        extendedLock = new int[lockSize + keySize * 2][lockSize + keySize * 2];
        for (int i = 0; i < lockSize; i++) {
            for (int j = 0; j < lockSize; j++) {
                extendedLock[i + keySize][j + keySize] = lock[i][j];
            }
        }

        int[][] rotatedKey = new int[keySize][keySize];
        for (int i = 0; i < keySize; i++) {
            for (int j = 0; j < keySize; j++) {
                rotatedKey[i][j] = key[i][j];
            }
        }

        for (int i = 0; i < 4; i++) {
            rotatedKey = rotate(rotatedKey);
            for (int startY = 0; startY < keySize + lockSize; startY++) {
                for (int startX = 0; startX < keySize + lockSize; startX++) {
                    int[][] combinedExtendedLock = combine(rotatedKey, startY, startX);
                    if (isRightKey(combinedExtendedLock)) {
                        return true;
                    }
                }
            }
        }

        return false;
    }
}